### DISCUSSION TITLE
"onLanguage:xxx" given more precedence than "contributions/languages/extensions" values in marketplace search with @ext:?

### DISCUSSION THREAD

**Discussion by tomqwpl** [microsoft/vscode-discussions#925] [Extension Development QnA] [answered]
Perhaps I misunderstand the "@ext:" search. However, if you open a file in vscode with an extension of "xxx" and try and set a language mode it gives you the option of searching the marketplace for extensions. The search it does is for "@ext:xxx". 

However, exploring the results that are returned, it appears that "onLanguage:xxx" is given precedence to extensions that actually declare a language contribution with an extension of ".xxx". This is based on the standard sort order of "relevance". That's got to be wrong hasn't it? The language ID in "onLanguage:xxx" could have nothing to do with the actual file extension ("python" is the language id for example, ".py" the extension). So it's actually unclear why this even has weight at all. For the "onLanguage:python" activation event to be fired, an extension has to have declared that ".py" is an extension associated with language ID "python", so you need the language mode extension before any others that might provide useful addons for that language. 

It's possible that there are other things at play, such as that number of downloads has a higher weight. However there are a number of extensions such as "Colorful comments" that declare "onLanguage:xxx" activation events for a whole bunch of language IDs but don't actually declare any language modes at all. These extensions would then be downloaded way more times because they are downloaded by users of a whole bunch of languages, rather than just one. These then swamp and hide the useful extensions that actually declare language modes for the file extension you were after.

**Comment by isidorn** [comment, marked as answer]
This is good feedback, and something for the Marketplace team to be aware of. We can not control the sort order from the VS Code side

fyi @seaniyer
